From 01e01feba6403c06c1f1cb5219a5134151a7fc80 Mon Sep 17 00:00:00 2001
From: Bharat Bhushan <Bharat.Bhushan@nxp.com>
Date: Tue, 7 Mar 2017 12:36:21 +0530
Subject: [PATCH 536/706] vfio fsl-mc: Reset container on _release()

dprc_reset_container() does complete cleanup and thus
allows re-run guest/user-space after abrupt guest kill.

Signed-off-by: Bharat Bhushan <Bharat.Bhushan@nxp.com>
(cherry picked from commit 10b1e3e3adad2516b5422432281444ca51a14c06)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/vfio/fsl-mc/vfio_fsl_mc.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/vfio/fsl-mc/vfio_fsl_mc.c b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
index 176d88b60de9..7b0b7dc8d39c 100644
--- a/drivers/vfio/fsl-mc/vfio_fsl_mc.c
+++ b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
@@ -110,6 +110,7 @@ static int vfio_fsl_mc_open(void *device_data)
 static void vfio_fsl_mc_release(void *device_data)
 {
 	struct vfio_fsl_mc_device *vdev = device_data;
+	struct fsl_mc_device *mc_dev = vdev->mc_dev;
 
 	mutex_lock(&driver_lock);
 
@@ -118,6 +119,10 @@ static void vfio_fsl_mc_release(void *device_data)
 		vfio_fsl_mc_irqs_cleanup(vdev);
 	}
 
+	if (strcmp(mc_dev->obj_desc.type, "dprc") == 0)
+		dprc_reset_container(mc_dev->mc_io, 0, mc_dev->mc_handle,
+				     mc_dev->obj_desc.id);
+
 	mutex_unlock(&driver_lock);
 
 	module_put(THIS_MODULE);
-- 
2.17.1

